cli/dcos-spark/submit_builder.go

```diff
@@ -280,75 +292,87 @@ func parseApplicationFile(args *sparkArgs) error {
+		// if current is a spark jar/app, we've processed all flags
+		case isSparkApp(current):
+			sparkArgs = append(sparkArgs, args[i])
+			appArgs = append(appArgs, args[i+1:]...)
+			break LOOP
+		case strings.HasPrefix(current, "--"):
+			isBool, skipNext := checkIfBooleanFlag(boolVals, args, i)
+			if isBool {
+				sparkArgs = append(sparkArgs, current)
+				i++
+				if skipNext {
+					i++
 				}
+				continue LOOP
 			}
-			break
-		}
-		// join this arg to the next arg if...:
-		// 1. we're not at the last arg in the array
-		// 2. we start with "--"
-		// 3. we don't already contain "=" (already joined)
-		// 4. we aren't a boolean value (no val to join)
-		if i < len(args)-1 && strings.HasPrefix(arg, "--") && !strings.Contains(arg, "=") {
-			// check for boolean:
-			for _, boolVal := range boolVals {
-				if boolVal.flagName == arg[2:] {
-					argsEquals = append(argsEquals, arg)
-					i += 1
-					continue ARGLOOP
-				}
+			if strings.Contains(current, "=") {
+				// already in the form arg=val, no merge required
+				sparkArgs = append(sparkArgs, current)
+				i++
+				continue LOOP
 			}
-			// merge this --key against the following val to get --key=val
-			argsEquals = append(argsEquals, arg+"="+args[i+1])
+			// otherwise, merge current with next into form arg=val; eg --driver-memory=512m
+			next := args[i+1]
+			sparkArgs = append(sparkArgs, current+"="+next)
 			i += 2
-		} else {
-			// already joined or at the end, pass through:
-			argsEquals = append(argsEquals, arg)
-			i += 1
+		default:
+			// if not a flag or jar, current is a continuation of the last arg and should not have been split
+			// eg extraJavaOptions="-Dparam1 -Dparam2" was parsed as [extraJavaOptions, -Dparam1, -Dparam2]
+			combined := sparkArgs[len(sparkArgs)-1] + " " + current
+			sparkArgs = append(sparkArgs[:len(sparkArgs)-1], combined)
+			i++
+		}
+	}
+	if config.Verbose {
+		client.PrintVerbose("Translated spark-submit arguments: '%s'", strings.Join(sparkArgs, ", "))
+		client.PrintVerbose("Translated application arguments: '%s'", strings.Join(appArgs, ", "))
+	}
+	return sparkArgs, appArgs
+}
+
+func isSparkApp(str string) bool {
+	for _, ext := range AcceptedSparkAppExtensions {
+		if strings.HasSuffix(str, ext) {
+			return true
 		}
 	}
-	client.PrintVerbose("Translated spark-submit arguments: '%s'", argsEquals)
-	client.PrintVerbose("Translated application arguments: '%s'", appFlags)
+	return false
+}
 
-	return argsEquals, appFlags
+func checkIfBooleanFlag(boolVals []*sparkVal, args []string, idx int) (isBool, skipNext bool) {
+	// check if current is a boolean flag (eg --supervise) but skip any accompanying boolean values (eg --supervise true)
+	current := args[idx]
+	for _, boolVal := range boolVals {
+		if boolVal.flagName == current[2:] {
+			isBool = true
+			next := args[idx+1]
+			if _, err := strconv.ParseBool(next); err == nil {
+				skipNext = true
```

Comment:
This is not correct. What if someone calls the function with `--supervise false` that should have different behaviour than `--supervise true`, but this is not captured here.

Also, as far as I know named return values have fallen out of favour in idiomatic Go.

Reply:
Ah... that's true. It may make sense to remove it from the list in the `false` case since that is default Spark behavior.

As for point 2, I didn't know. I was just operating off the Go reference and thought it was a nifty feature.

I did some googling and found this as the latest reference but it was in favor of named return values: https://blog.minio.io/golang-internals-part-2-nice-benefits-of-named-return-values-1e95305c8687

Can you point me to resources that argue the opposite?

Reply:
I may have just been recalling a recording of a gophercon talk that I watched. In this case, I don't think it's the named return values so much as the naked return that is the issue. I stand corrected.

Here is a PR review that makes the changes to the Go standard library and style guide: 
* https://go-review.googlesource.com/c/go/+/20024 
* https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

For clarity here:
> Named returned values should only be used on public funcs and methods
> when it contributes to the documentation.
>
> Named return values should not be used if they're only saving the
> programmer a few lines of code inside the body of the function,
> especially if that means there's stutter in the documentation or it
> was only there so the programmer could use a naked return
> statement. (Naked returns should not be used except in very small
> functions)

in this case, I would say that we are not stuttering so the named values themselves are ok, but we should consider dropping the naked return -- although we could consider this a very small function.